rust/connlib/libs/tunnel/src/device_channel_unix.rs

```diff
@@ -1,74 +1,78 @@
-use std::sync::Arc;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering::Relaxed},
+    Arc,
+};
 
-use libs_common::{messages::Interface, CallbackErrorFacade, Callbacks, Error, Result};
+use ip_network::IpNetwork;
+use libs_common::{messages::Interface, CallbackErrorFacade, Callbacks, Result};
 use tokio::io::unix::AsyncFd;
 
-use crate::tun::{IfaceConfig, IfaceDevice};
+use crate::tun::{IfaceDevice, IfaceStream};
 
-#[derive(Debug)]
-pub(crate) struct DeviceChannel(AsyncFd<Arc<IfaceDevice>>);
+pub(crate) struct IfaceConfig {
+    mtu: AtomicUsize,
+    iface: IfaceDevice,
+}
 
-impl DeviceChannel {
-    pub(crate) async fn mtu(&self) -> Result<usize> {
-        self.0.get_ref().mtu().await
-    }
+#[derive(Clone)]
+pub(crate) struct DeviceIo(Arc<AsyncFd<IfaceStream>>);
 
-    pub(crate) async fn read(&self, out: &mut [u8]) -> std::io::Result<usize> {
+impl DeviceIo {
+    pub async fn read(&self, out: &mut [u8]) -> std::io::Result<usize> {
         loop {
             let mut guard = self.0.readable().await?;
 
-            match guard.try_io(|inner| {
-                inner.get_ref().read(out).map_err(|err| match err {
-                    Error::IfaceRead(e) => e,
-                    _ => panic!("Unexpected error while trying to read network interface"),
-                })
-            }) {
-                Ok(result) => break result.map(|e| e.len()),
+            match guard.try_io(|inner| inner.get_ref().read(out)) {
```

Comment:
In recent versions of tokio, you can use `async_io`: https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html#method.async_io